lib/kang.js

```diff
@@ -80,20 +85,42 @@ function knMakeSource(str)
 		str = str.substr(0, c);
 	}
 
-	c = str.indexOf(':');
-	if (c == -1) {
-		host = str;
+	if ((matched = IPV4_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
+		if (!mod_net.isIPv4(host)) {
+			throw new Error('bad IPv4 address: ' +
+			    JSON.stringify(host));
+		}
+	} else if ((matched = IPV6_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
+		if (!mod_net.isIPv6(host)) {
+			throw new Error('bad IPv6 address: ' +
+			    JSON.stringify(host));
+		}
+	} else if ((matched = DOMAIN_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
 	} else {
-		portstr = str.substr(c + 1);
+		throw (new Error('no host specified'));
+	}
+
+	if (portstr !== undefined) {
+		mod_assert.equal(portstr[0], ':');
+		portstr = portstr.slice(1);
 		port = parseInt(portstr, 10);
-		host = str.substr(0, c);
 
-		if (isNaN(port))
-			throw (new Error('invalid port: "' + portstr + '"'));
-	}
+		if (Number(portstr) !== Math.floor(port)) {
```

Comment:
If we're going to be this precise about it (which isn't a bad idea), we should probably add a proper function to jsprim to do this.  I've been meaning to do that for a while.  That's not a blocker here but my bias has been to avoid adding non-trivial and likely differing logic to multiple different consumers in the absence of such a function.

Which cases does this check cover?  Is it that the port string contains non-integer trailing characters (like "3.2" or "3q")?